apps/agentic-chat/src/hooks/useTools.ts

```diff
@@ -0,0 +1,466 @@
+import { useAccount, useWalletClient } from 'wagmi';
+import {
+  Address,
+  Chain,
+  createPublicClient,
+  encodeFunctionData,
+  erc20Abi,
+  getAddress,
+  Hex,
+  http,
+  PublicClient,
+} from 'viem';
+import { ToolCall } from '@ai-sdk/provider-utils';
+import axios from 'axios';
+import qs from 'qs';
+import {
+  BebopResponse,
+  BebopQuote,
+  Asset,
+  PortalsResponse,
+  PortalsToken,
+} from '@agentic-chat/types';
+
+import {
+  arbitrum,
+  mainnet,
+  polygon,
+  optimism,
+  base,
+  avalanche,
+  bsc,
+  gnosis,
+} from 'viem/chains';
+import {
+  ASSET_NAMESPACE,
+  AssetId,
+  CHAIN_NAMESPACE,
+  ChainId,
+} from '@shapeshiftoss/caip';
+import { fromBaseUnit, toBaseUnit } from '@agentic-chat/utils';
+import { useState } from 'react';
+
+const BEBOP_ETH_MARKER = '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE';
+
+const getChainById = (chainId: number): Chain => {
+  const chains: Record<number, Chain> = {
+    [mainnet.id]: mainnet,
+    [arbitrum.id]: arbitrum,
+    [polygon.id]: polygon,
+    [optimism.id]: optimism,
+    [base.id]: base,
+    [avalanche.id]: avalanche,
+    [gnosis.id]: gnosis,
+    [bsc.id]: bsc,
+  };
+  const chain = chains[chainId];
+  if (!chain) throw new Error(`Unsupported chain ID: ${chainId}`);
+  return chain;
+};
+
+const getPublicClient = (chainId: number): PublicClient => {
+  const chain = getChainById(chainId);
+  return createPublicClient({
+    chain,
+    transport: http(),
+  });
+};
+
+const env = import.meta?.env ? import.meta.env : process.env;
+
+const PORTALS_BASE_URL = env.VITE_PORTALS_BASE_URL;
+const PORTALS_API_KEY = env.VITE_PORTALS_API_KEY;
+const BEBOP_API_KEY = env.VITE_BEBOP_API_KEY;
+
+const useTools = () => {
+  const account = useAccount();
+  const { data: walletClient } = useWalletClient();
+  const [bebopQuote, setBebopQuote] = useState<BebopQuote | null>(null);
+
+  const handleToolCall = async ({
+    toolCall,
+  }: {
+    toolCall: ToolCall<string, unknown>;
+  }) => {
+    if (!PORTALS_BASE_URL || !PORTALS_API_KEY || !BEBOP_API_KEY) {
+      throw new Error('Missing env vars');
+    }
+
+    switch (toolCall.toolName) {
```

Comment:
Fine for now, but we'll want to tidy this up a bit - i.e. extract each tool into its own fn and then just call that fn in the switch.